### DISCUSSION TITLE
new parquet reader failed, err=parquet: could not retrieve footer offset: The specified key does not exist.: importing data failed

### DISCUSSION THREAD

**Discussion by WanBenLe** [milvus-io/milvus#36203] [Q&A and General discussion]
from pymilvus import MilvusClient, DataType
from tqdm.auto import tqdm
from pymilvus import Collection, utility
from pymilvus.bulk_writer import LocalBulkWriter, BulkFileType

client = MilvusClient(uri="http://localhost:19530")
if client.has_collection(collection_name="demo_collection"):
    client.drop_collection(collection_name="demo_collection")
schema = client.create_schema(
    auto_id=False,
    enable_dynamic_field=False
)
schema.add_field(field_name="sku", datatype=DataType.VARCHAR, is_primary=True, max_length=128)
schema.add_field(field_name="vec", datatype=DataType.FLOAT_VECTOR, dim=768)
schema.add_field(field_name="title", datatype=DataType.VARCHAR, max_length=512)
schema.verify()
index_params = MilvusClient.prepare_index_params()
index_params.add_index(field_name="vec",index_type="HNSW",metric_type="IP",efConstruction=40, M= 512)
client.create_collection(collection_name="demo_collection",index_params=index_params,schema=schema)
writer = LocalBulkWriter( schema=schema,local_path='.',segment_size=5*1024 * 1024 * 1024, file_type=BulkFileType.PARQUET)
for i in tqdm(range(len(sku))):
    writer.append_row({"sku": str(sku[i])[:128], "vec":vectors[i],"title": str(title[i])[:512]})
writer.commit()

import os
os.system("""
export MILVUS_URI="localhost:19530"

curl --request POST "http://${MILVUS_URI}/v2/vectordb/jobs/import/create" \
--header "Content-Type: application/json" \
--data-raw '{
    "files": [
        [
           "/data/37b0f2b9-96bd-4a08-b99c-51a94b36f1db/1.parquet"
        ]
    ],
    "collectionName": "demo_collection"
}'
"""
)

os.system("""
export MILVUS_URI="localhost:19530"
curl --request POST "http://${MILVUS_URI}/v2/vectordb/jobs/import/get_progress" \
--header "Content-Type: application/json" \
--data-raw '{
    "jobId": "~"
}'
""")

will raise

{"code":0,"data":{"collectionName":"demo_collection","completeTime":"","details":[],"fileSize":0,"importedRows":0,"jobId":"452491787553486304","progress":0,"reason":"new parquet reader failed, err=parquet: could not retrieve footer offset: The specified key does not exist.: importing data failed","state":"Failed","totalRows":0}}

how can I fix it?

**Comment by yhmo** [comment]
The **LocalBulkWriter only generates a parquet file in the local disk, while the bulk_insert requires the file to be uploaded to the minio/S3**.
You can use RemoteBulkWriter to generate a parquet file and upload the file to minio/S3. You need to provide the access parameters to the minio/s3 for the RemoteBulkWriter

```
from pymilvus import MilvusClient, DataType
import random
from pymilvus.bulk_writer import RemoteBulkWriter, BulkFileType


client = MilvusClient(uri="http://localhost:19530")
if client.has_collection(collection_name="demo_collection"):
    client.drop_collection(collection_name="demo_collection")

schema = client.create_schema(
    auto_id=False,
    enable_dynamic_field=False
)
schema.add_field(field_name="sku", datatype=DataType.VARCHAR, is_primary=True, max_length=128)
schema.add_field(field_name="vec", datatype=DataType.FLOAT_VECTOR, dim=768)
schema.add_field(field_name="title", datatype=DataType.VARCHAR, max_length=512)
schema.verify()

index_params = MilvusClient.prepare_index_params()
index_params.add_index(field_name="vec",index_type="HNSW",metric_type="IP",efConstruction=40, M= 512)
client.create_collection(collection_name="demo_collection",index_params=index_params,schema=schema)

MINIO_ADDRESS = "0.0.0.0:9000"
MINIO_SECRET_KEY = "minioadmin"
MINIO_ACCESS_KEY = "minioadmin"
writer = RemoteBulkWriter(
    schema=schema,
    remote_path="bulk_data",
    connect_param=RemoteBulkWriter.S3ConnectParam(
                endpoint=MINIO_ADDRESS,
                access_key=MINIO_ACCESS_KEY,
                secret_key=MINIO_SECRET_KEY,
                bucket_name="a-bucket",
            ),
    segment_size=5*1024 * 1024 * 1024,
    file_type=BulkFileType.PARQUET
)
for i in range(100):
    writer.append_row({"sku": f"No_{i}", "vec":[random.random() for _ in range(768)],"title": f"title_{i}"})
writer.commit()

print(writer.batch_files)

```

Once the RemoteBulkWriter is committed, the "writer.batch_files" is the remote path of the parquet file on the minio/s3, you can use this remote file to call bulk_insert interface:

If you run the script, you will get a remote path like this:
`[['bulk_data/b0a693c1-7a3d-4554-a702-3b0a76d5dfdc/1.parquet']]`

Then you can use this path to call bulk_insert:

```
export MILVUS_URI="localhost:19530"

curl --location --request POST "http://${MILVUS_URI}/v2/vectordb/jobs/import/create" \
--header "Content-Type: application/json" \
--data-raw '{
    "files": [
        [
            "bulk_data/b0a693c1-7a3d-4554-a702-3b0a76d5dfdc/1.parquet"
        ]
    ],
    "collectionName": "demo_collection"
}'
```
